### PR TITLE
Switched AST benchmarks from GoogleBench to NVBench

### PR DESCRIPTION
**Pull request by lamarrr**
## Description
This merge request switches the Benchmarking solution for the AST benchmark from GoogleBench to NVBench. ~It also refactors the L2 cache flushing functionality of `cuda_event_timer` into a separate function `flush_L2_device_cache`, since NVBench already performs the timing, synchronization, and timer setup necessary.~

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/benchmarks/ast/transform.cpp [hunk 1]
```diff
@@ -15,14 +15,18 @@
...
 #include <benchmarks/synchronization/synchronization.hpp>
 
 #include <cudf/transform.hpp>
 #include <cudf/types.hpp>
 
+#include <rmm/cuda_stream_view.hpp>
+
 #include <thrust/iterator/counting_iterator.h>
 
+#include <nvbench/exec_tag.cuh>
```
Comment (karthikeyann) [inline cpp/benchmarks/ast/transform.cpp:27]: nit: This include is already included in `<nvbench/nvbench.cuh>`. so, this include may not be required.

File: cpp/benchmarks/ast/transform.cpp [hunk 2]
```diff
@@ -86,38 +87,25 @@ static void BM_ast_transform(benchmark::State& state)
...
+    ::BM_ast_transform<key_type, tree_type, reuse_columns, nullable>(st);                  \
   }                                                                                        \
-  BENCHMARK_REGISTER_F(AST, name)                                                          \
-    ->Apply(CustomRanges)                                                                  \
-    ->Unit(benchmark::kMillisecond)                                                        \
-    ->UseManualTime();
+  NVBENCH_BENCH(name)                                                                      \
+    .set_name(#name)                                                                       \
+    .add_int64_axis("TableSize", {100'000, 1'000'000, 10'000'000, 100'000'000})            \
+    .add_int64_axis("TreeLevels", {1, 5, 10})
```
Comment (karthikeyann) [inline cpp/benchmarks/ast/transform.cpp:108]: nit: please check if other nvbench benchmarks use CamelCase or snake_case for axis names and align to that. AFAIK, nvbench uses snake_case.

**Review by karthikeyann — COMMENTED** [review]
clean porting. appreciate the minimal diff.

File: cpp/benchmarks/ast/transform.cpp (hunk 1 above)
Comment (lamarrr) [reply cpp/benchmarks/ast/transform.cpp:27]: thanks!

File: cpp/benchmarks/ast/transform.cpp (hunk 2 above)
Comment (bdice) [reply cpp/benchmarks/ast/transform.cpp]: Good question -- thanks for asking that. I am not sure off the top of my head.

**Review by bdice — APPROVED** [review]
This is nice, thanks!

File: cpp/benchmarks/synchronization/synchronization.cpp [hunk 3]
```diff
@@ -21,26 +21,29 @@
 #include <rmm/cuda_stream_view.hpp>
 #include <rmm/device_buffer.hpp>
 
+void flush_device_L2_cache(rmm::cuda_stream_view stream)
```
Comment (PointKernel) [inline cpp/benchmarks/synchronization/synchronization.cpp:24]: nvbench internally flushes l2 by default (https://github.com/NVIDIA/nvbench/blob/a171514056e5d6a7f52a035dd6c812fa301d4f4f/nvbench/detail/measure_cold.cuh#L120). No need to port this
```suggestion
```

**Review by PointKernel — CHANGES_REQUESTED** [review]

**Comment by bdice** [comment]
@lamarrr Can you avoid force-pushing? It's helpful for reviewers when we can see the history. I just reviewed as you pushed, and now I don't know how to easily compare the state before and after your push. Generally, force-pushing is a last resort for us, only if it's strictly needed.

File: cpp/benchmarks/synchronization/synchronization.cpp (hunk 3 above)
Comment (lamarrr) [reply cpp/benchmarks/synchronization/synchronization.cpp:24]: Awesome! thanks!

**Comment by lamarrr** [comment]
> @lamarrr Can you avoid force-pushing? It's helpful for reviewers when we can see the history. I just reviewed as you pushed, and now I don't know how to easily compare the state before and after your push. Generally, force-pushing is a last resort for us, only if it's strictly needed.

noted!

File: cpp/benchmarks/CMakeLists.txt [hunk 4]
```diff
@@ -330,7 +330,7 @@ ConfigureNVBench(CSV_WRITER_NVBENCH io/csv/csv_writer.cpp)
 
 # ##################################################################################################
 # * ast benchmark ---------------------------------------------------------------------------------
-ConfigureBench(AST_BENCH ast/transform.cpp)
+ConfigureNVBench(AST_BENCH ast/transform.cpp)
```
Comment (PointKernel) [inline cpp/benchmarks/CMakeLists.txt:333]: IIRC, nightly benchmarks rely on the proper suffix to distinguish between nvbench and gbench so we may want to update this to `AST_NVBENCH` as well.

Comment (lamarrr) [reply cpp/benchmarks/CMakeLists.txt:333]: alright

File: cpp/benchmarks/ast/transform.cpp [hunk 5]
```diff
@@ -86,38 +85,23 @@ static void BM_ast_transform(benchmark::State& state)
...
-{
-  auto row_counts       = std::vector<cudf::size_type>{100'000, 1'000'000, 10'000'000, 100'000'000};
-  auto operation_counts = std::vector<cudf::size_type>{1, 5, 10};
-  for (auto const& row_count : row_counts) {
-    for (auto const& operation_count : operation_counts) {
-      b->Args({row_count, operation_count});
-    }
-  }
+  state.add_global_memory_reads(static_cast<int64_t>(state.get_summaries().size()) * table_size *
+                                (tree_levels + 1) * sizeof(key_type));
```
Comment (PointKernel) [inline cpp/benchmarks/ast/transform.cpp:93]:
```suggestion
  state.add_global_memory_reads<key_type>(table_size * (tree_levels + 1));
```
Nvbench and gbench compute memory bandwidth in a slightly different way, so there's no need to manually calculate the number of bytes. After making this change, could you please double-check to ensure that the gbench results align coherently with the nvbench ones?

for your reference: https://github.com/NVIDIA/nvbench/blob/e8c8877d3669e30e977cf6258d96e1c645922077/examples/throughput.cu

**Review by PointKernel — APPROVED** [review]
LGTM

Thanks for the quick turnaround and addressing all my review comments!

**Comment by lamarrr** [comment]
> LGTM
> 
> Thanks for the quick turnaround and addressing all my review comments!

thank you

File: cpp/benchmarks/ast/transform.cpp [hunk 6]
```diff
@@ -86,38 +85,22 @@ static void BM_ast_transform(benchmark::State& state)
...
-  for (auto _ : state) {
-    cuda_event_timer raii(state, true);  // flush_l2_cache = true, stream = 0
-    cudf::compute_column(table, expression_tree_root);
-  }
-
   // Use the number of bytes read from global memory
-  state.SetBytesProcessed(static_cast<int64_t>(state.iterations()) * state.range(0) *
-                          (tree_levels + 1) * sizeof(key_type));
-}
+  state.add_global_memory_reads<key_type>(table_size * (tree_levels + 1));
```
Comment (ttnghia) [inline cpp/benchmarks/ast/transform.cpp:89]: Wrong number of bytes added. Need to multiply with `sizeof(key_type)`.

**Review by ttnghia — APPROVED** [review]

**Pull request merged**